libraries/SITL/SIM_Aircraft.cpp

```diff
@@ -258,31 +253,29 @@ void Aircraft::sync_frame_time(void)
 {
     frame_counter++;
     uint64_t now = get_wall_time_us();
-    if (frame_counter >= 40 &&
-        now > last_wall_time_us) {
-        const float rate = frame_counter * 1.0e6f/(now - last_wall_time_us);
-        achieved_rate_hz = (0.99f*achieved_rate_hz) + (0.01f * rate);
-        if (achieved_rate_hz < rate_hz * target_speedup) {
-            scaled_frame_time_us *= 0.999f;
-        } else {
-            scaled_frame_time_us /= 0.999f;
-        }
-#if 0
-        ::printf("achieved_rate_hz=%.3f rate=%.2f rate_hz=%.3f sft=%.1f\n",
-                 static_cast<double>(achieved_rate_hz),
-                 static_cast<double>(rate),
-                 static_cast<double>(rate_hz),
-                 static_cast<double>(scaled_frame_time_us));
-#endif
-        const uint32_t sleep_time = static_cast<uint32_t>(scaled_frame_time_us * frame_counter);
-        if (sleep_time > min_sleep_time) {
-            usleep(sleep_time);
-        }
-        last_wall_time_us = now;
-        frame_counter = 0;
+    uint64_t dt_us = now - last_wall_time_us;
+
+    const float target_dt_us = 1.0e6/(rate_hz*target_speedup);
+
+    // accumulate sleep debt if we're running too fast
+    sleep_debt_us += target_dt_us - dt_us;
+
+    if (sleep_debt_us < -1.0e5) {
+        // don't let a large negative debt build up
+        sleep_debt_us = -1.0e5;
     }
-    uint32_t now_ms = now / 1000ULL;
-    if (now_ms - last_fps_report_ms > 2000) {
+    if (sleep_debt_us > min_sleep_time) {
```

Comment:
shouldn't the sleep time caped by the target_dt_us in order to not oversleep too much ? after a sleep the sleep_debt_us will mostly always be at max negative

Reply:
> shouldn't the sleep time caped by the target_dt_us in order to not oversleep too much ?

no, we normally will sleep much more than target_dt_us

> after a sleep the sleep_debt_us will mostly always be at max negative
why? It will normally be close to zero if the sleep was accurate